drivers/rilmodem/gprs-context.c

```diff
@@ -534,9 +544,27 @@ static void ril_gprs_context_remove(struct ofono_gprs_context *gc)
 	DBG("*gc: %p", gc);
 
 	if (gcd->state != STATE_IDLE) {
-		ril_gprs_context_detach_shutdown(gc, 0);
+		struct req_deactivate_data_call request;
+		struct parcel rilp;
+		struct ofono_error error;
+
+		request.cid = gcd->active_rild_cid;
+		request.reason = RIL_DEACTIVATE_DATA_CALL_NO_REASON;
+		g_ril_request_deactivate_data_call(gcd->ril, &request,
+								&rilp, &error);
+
+		g_ril_send(gcd->ril, RIL_REQUEST_DEACTIVATE_DATA_CALL,
+						&rilp, NULL, NULL, NULL);
+	}
+
```

Comment:
Going through your points:
* The state check is already performed in ril_gprs_context_remove ("if (gcd->state != STATE_IDLE)" in line 546. The only difference is that the deactivate call request is sent also if state is DISABLING, but I think that is fine (would be kind of a last try before the atom gets removed).
* The unregister is performed anyway when g_ril_unref() is called right after sending the request.
* The retries would not be performed anywayt, because g_ril_unref() will remove any pending callbacks after being called. Which is good, as we are removing the atom in the modem logic right after ril_gprs_context_remove() is called.

Note that I think the crash happens when we are re-trying to close the connection and the atom gets destroyed with pending callbacks around introduced by g_timeout_add_seconds().